Add requirements for built docs to new package guide (#554)
lsst-dm/dm_dev_guide: stack/adding-a-new-package.rst

If the new package needs a distinct Jira component (most will), any DMLT member (such as your manager) can add one.

Adding a package to a distributed product
=========================================

RFC process
-----------

New packages intended for distribution to end users should generally be added as a dependency of a "top-level product:" these are the roots of the LSST package hierarchy.
They include ``lsst_apps``, ``lsst_distrib``, ``qserv_distrib`` and ``lsst_sims``.

> Before adding a new dependency to one these top-level distribution products, it must be approved through the :doc:`RFC process </communications/rfc>`.

Reviewer: ```suggestion
Before adding a new dependency to one of these top-level distribution products, it must be approved through the :doc:`RFC process </communications/rfc>`.
```

Consensus must be reached regarding both the name and the suitability of the new package.
Before adopting the RFC, the following steps must be completed:

* Implementation tickets are created to cover package creation.
* The package is migrated to the `LSST organization on GitHub`_, if not already there.

> * The package must have a descriptive readme file at the root level, describing how it fits into the Science Pipelines infrastructure.

Reviewer: ```suggestion
* The package must have a descriptive README file at the root level, describing how it fits into the Science Pipelines infrastructure.
```

> * The package docs must build and be linked from a branch of the `pipelines docs`_.

Reviewer 2: ```suggestion
* The package docs must build and be linked from a branch of the `Science Pipelines docs`_.
```

> * The package docs must build and be linked from a branch of the `Science Pipelines docs`_.

Reviewer: A package RFC can be adopted before any code exists.  How can we insist that the docs must already exist?

I think the most you can do is say that this must be part of the implementation tickets mentioned in the first bullet, and that the docs must build prior to marking the RFC as implemented.

Author: Can a package RFC for "add this to lsst_distrib" really be adopted before any code exists? Can it even really be written, since you can't say anything about how it actually connects to the rest of the ecosystem. Even if that's theoretically possible, I'm not aware of us ever done an "add a package that hasn't been created yet to the build" RFC.

Reviewer: You can say what it's intended to do.  As a sort-of example, the drp_pipe RFC (775) didn't exactly have code for all of the packages before it was adopted.

  The automated builds use the team membership to determine the type of tag to be applied.
  Having the code reside in the ``lsst`` or ``lsst-dm`` organization on GitHub is not sufficient.

repos.yaml
----------

The new package must be added to the `etc/repos.yaml file in the lsst/repos repository`_ along with its corresponding GitHub URL.
This file is governed by a "self-merge" policy: upon opening a pull request, it will be checked by GitHub Actions, and developers may merge without further review on success.
This change **must** be merged before the package can be built on Jenkins, and this should be done prior to the RFC being filed.
Refer to :jira:`RFC-75` for background.

Package and pipelines docs
--------------------------

Documentation for the new package must be built and linked from the main `pipelines docs`_ page.
To make that documentation visible as part of your RFC before your package is included in the official builds:

1. Follow the instructions for :ref:`adding a package to pipelines.lsst.io <add-to-pipelines-lsst-io>` on a ticket branch.

> 2. :ref:`Build the pipelines docs locally <local-pipelines-lsst-io-build>` on that branch.

Reviewer 2: ```suggestion
2. :ref:`Build the Science PIpelines docs locally <local-pipelines-lsst-io-build>` on that branch.
```

3. Copy the ``_build/html`` directory from your pipelines build to a place that's publicly viewable (e.g. your public web path on :doc:`the USDF </usdf/storage>`).
4. Include a link to those built docs in your RFC.

> Unfortunately, our current pipelines docs build system only allows building against tagged versions of packages (e.g. weekly tags), and your new package cannot be tagged until it is added as a dependency (see below).

Reviewer 2: ```suggestion
.. note::

   The current Science Pipelines documentation build only uses building against tagged versions of packages (e.g. daily or weekly tags) in a release with a Docker image build, like ``lsst_distrib``.
   You can work around this current limitation by building the documentation locally and publishing it with your USDF web hosting, as described above.
```

Reviewer 2: (I think Docker images are also build for daily releases)

> This is why you have to do a local build with your branch and put it somewhere to it be viewed by those reading the RFC, instead of using our :ref:`Jenkins system <jenkins-pipelines-lsst-io-build>`.

Reviewer 2: Delete (see above)

  The automated builds use the team membership to determine the type of tag to be applied.
  Having the code reside in the ``lsst`` or ``lsst-dm`` organization on GitHub is not sufficient.

repos.yaml
----------

The new package must be added to the `etc/repos.yaml file in the lsst/repos repository`_ along with its corresponding GitHub URL.
This file is governed by a "self-merge" policy: upon opening a pull request, it will be checked by GitHub Actions, and developers may merge without further review on success.

> This change **must** be merged before the package can be built on Jenkins, and this should be done prior to the RFC being filed.

Reviewer: ```suggestion
This change **must** be merged before the package can be built on Jenkins, and this should be done prior to the RFC being adopted.
```

Choosing the package name is part of the RFC process, so the name can't be used prior to filing the RFC.  Adoption is the earliest gate that could be used, but I could see an argument for not doing repos.yaml until final implementation.

Author: I changed it to "early in the implementation process", since it's a prereq for a lot of other things.

Refer to :jira:`RFC-75` for background.

Package and pipelines docs
--------------------------

Documentation for the new package must be built and linked from the main `Science Pipelines docs`_ page.
To make that documentation visible as part of your RFC before your package is included in the official builds:

1. Follow the instructions for :ref:`adding a package to pipelines.lsst.io <add-to-pipelines-lsst-io>` on a ticket branch.
2. :ref:`Build the Science PIpelines docs locally <local-pipelines-lsst-io-build>` on that branch.

> 3. Copy the ``_build/html`` directory from your pipelines build to a place that's publicly viewable (e.g. your public web path on :doc:`the USDF </usdf/storage>`).

Reviewer 3: I don't believe there is any public web space on the USDF at this point (e.g., [this Slack thread](https://lsstc.slack.com/archives/C027902HL8H/p1661880029064899)).  So I suggest deleting the parenthetical for the time being.

Author: My thought with this is that we'll have some kind of usable public web space there soon enough, and we won't remember to update this with that suggestion once we do (and I expect we'll have public web space before someone follows this guide, but maybe with gbdes incoming that's not true?).

Author: I've updated the link to point to the new instructions for setting up a public page.

> 4. Include a link to those built docs in your RFC.

Reviewer: Again, I think this is premature.  The RFC happens at the beginning of the process, not the end.

Author: I've added some wording about RFC timing: please let me know what you think.

.. _repos.yaml: https://github.com/lsst/repos/blob/main/etc/repos.yaml
.. _manifest.remap:  https://github.com/lsst/lsstsw/blob/main/etc/manifest.remap
.. _pipe_base: https://github.com/lsst/pipe_base/

> .. _pipelines docs: https://pipelines.lsst.io/

Reviewer 2: ```suggestion
.. _Science Pipelines docs: https://pipelines.lsst.io/
```
